### PR TITLE
Implement drawing shadows in canvas.

### PR DESCRIPTION
**Pull request by hyowon**
https://html.spec.whatwg.org/multipage/#when-shadows-are-drawn
r? @nox 
cc @mmatyas @yichoi

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/6355)

### CODE REVIEW

**Comment by nox** [comment]
```
$ ./mach test-tidy
components/canvas/canvas_paint_task.rs:338: (much) overlong line
components/canvas/canvas_paint_task.rs:349: (much) overlong line
components/canvas/canvas_paint_task.rs:367: (much) overlong line
components/canvas/canvas_paint_task.rs:379: (much) overlong line
components/canvas/canvas_paint_task.rs:601: (much) overlong line
```

File: components/canvas/canvas_paint_task.rs [hunk 1]
```diff
@@ -301,8 +250,25 @@ impl<'a> CanvasPaintTask<'a> {
...
+            new_draw_target.fill_rect(rect, self.state.fill_style.to_pattern_ref(),
+                                      Some(&self.state.draw_options));
+
+            self.drawtarget.draw_surface_with_shadow(new_draw_target.snapshot(),
+                                                     &Point2D(source_rect.origin.x as AzFloat,
+                                                              source_rect.origin.y as AzFloat),
+                                                     &self.state.shadow_color,
+                                                     &Point2D(self.state.shadow_offset_x as AzFloat,
+                                                              self.state.shadow_offset_y as AzFloat),
+                                                     (self.state.shadow_blur / 2.0f64) as AzFloat,
```
Comment (pcwalton) [inline components/canvas/canvas_paint_task.rs]: Why divide by two here?

File: components/canvas/canvas_paint_task.rs [hunk 2]
```diff
@@ -361,7 +327,29 @@ impl<'a> CanvasPaintTask<'a> {
...
+            write_image(&new_draw_target, image_data, source_rect.size, dest_rect,
+                        smoothing_enabled, self.state.draw_options.alpha);
+
+            self.drawtarget.draw_surface_with_shadow(new_draw_target.snapshot(),
+                                                     &Point2D(shadow_src_rect.origin.x as AzFloat,
+                                                              shadow_src_rect.origin.y as AzFloat),
+                                                     &self.state.shadow_color,
+                                                     &Point2D(self.state.shadow_offset_x as AzFloat,
+                                                              self.state.shadow_offset_y as AzFloat),
+                                                     (self.state.shadow_blur / 2.0f64) as AzFloat,
```
Comment (pcwalton) [inline components/canvas/canvas_paint_task.rs]: Same here.

File: components/canvas/canvas_paint_task.rs (hunk 1 above)
Comment (hyowon) [inline components/canvas/canvas_paint_task.rs]: The shadowBlur attribute specifies the level of the blurring effect.
And the sigma as the standard deviation to perform a 2D Gaussian blur is half the value of shadowBlur.
Please refer to the 4-1 on the spec https://html.spec.whatwg.org/multipage/#when-shadows-are-drawn.

File: components/canvas/canvas_paint_task.rs (hunk 2 above)
Comment (hyowon) [inline components/canvas/canvas_paint_task.rs]: The same as above. :)

**Comment by nox** [comment]
I think draw_rect(), draw_image() and draw_image_self() should call a higher-order function draw_with_shadow() taking a closure and shadow_src_rect.

-S-awaiting-review +S-needs-code-changes 

---

Review status: all files reviewed, 5 unresolved discussions, all commit checks successful.
Reviewed files:
- components/canvas/canvas_paint_task.rs @ r3
- tests/wpt/metadata/2dcontext/drawing-rectangles-to-the-canvas/2d.fillRect.shadow.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.alpha.2.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.alpha.3.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.alpha.4.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.alpha.5.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.canvas.alpha.html.ini @ r2
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.canvas.basic.html.ini @ r2
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.canvas.transparent.2.html.ini @ r2
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.clip.1.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.clip.3.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.composite.1.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.composite.2.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.gradient.alpha.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.gradient.basic.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.gradient.transparent.2.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.image.alpha.html.ini @ r2
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.image.basic.html.ini @ r2
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.image.scale.html.ini @ r2
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.image.transparent.2.html.ini @ r2
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.offset.negativeX.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.offset.negativeY.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.offset.positiveX.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.offset.positiveY.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.outside.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.transform.1.html.ini @ r1
- tests/wpt/metadata/2dcontext/shadows/2d.shadow.transform.2.html.ini @ r1

---

<sup>**[components/canvas/canvas_paint_task.rs, line 625 \[r3\]](https://reviewable.io:443/reviews/servo/servo/6355#-JredawfidXyjcr11wQq)** ([raw file](https://github.com/servo/servo/blob/49b2a474fc0c791d5aba2c3fdf072f85efff09d4/components/canvas/canvas_paint_task.rs#L625)):</sup>
Link to spec.

https://html.spec.whatwg.org/multipage/scripting.html#when-shadows-are-drawn

---

<sup>**[components/canvas/canvas_paint_task.rs, line 626 \[r3\]](https://reviewable.io:443/reviews/servo/servo/6355#-JredeLtFX8Ix9H30SxR)** ([raw file](https://github.com/servo/servo/blob/49b2a474fc0c791d5aba2c3fdf072f85efff09d4/components/canvas/canvas_paint_task.rs#L626)):</sup>
The spec says "non-zero".

---

<sup>**[components/canvas/canvas_paint_task.rs, line 629 \[r3\]](https://reviewable.io:443/reviews/servo/servo/6355#-JredgVK0ICJ3CBJOS9m)** ([raw file](https://github.com/servo/servo/blob/49b2a474fc0c791d5aba2c3fdf072f85efff09d4/components/canvas/canvas_paint_task.rs#L629)):</sup>
The spec says that alpha should be non-zero and one of these three too. It lacks parentheses.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6355)

**Comment by hyowon** [comment]
As you commented, I've added draw_with_shadow(). It makes the code more concise. Thanks.

---

Review status: 26 of 27 files reviewed, 5 unresolved discussions, some commit checks failed.

---

<sup>**[components/canvas/canvas_paint_task.rs, line 625 \[r3\]](https://reviewable.io:443/reviews/servo/servo/6355#-JredawfidXyjcr11wQq)** ([raw file](https://github.com/servo/servo/blob/49b2a474fc0c791d5aba2c3fdf072f85efff09d4/components/canvas/canvas_paint_task.rs#L625)):</sup>
Done.

---

<sup>**[components/canvas/canvas_paint_task.rs, line 626 \[r3\]](https://reviewable.io:443/reviews/servo/servo/6355#-JredeLtFX8Ix9H30SxR)** ([raw file](https://github.com/servo/servo/blob/49b2a474fc0c791d5aba2c3fdf072f85efff09d4/components/canvas/canvas_paint_task.rs#L626)):</sup>
Done.

---

<sup>**[components/canvas/canvas_paint_task.rs, line 629 \[r3\]](https://reviewable.io:443/reviews/servo/servo/6355#-JredgVK0ICJ3CBJOS9m)** ([raw file](https://github.com/servo/servo/blob/49b2a474fc0c791d5aba2c3fdf072f85efff09d4/components/canvas/canvas_paint_task.rs#L629)):</sup>
Added parentheses. Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6355)

**Comment by nox** [comment]
-S-awaiting-review -S-needs-rebase +S-needs-squash 

Looks good, squash and I will r+ it.

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6355)

**Comment by nox** [comment]
@bors-servo: r+

-S-awaiting-review -S-needs-squash +S-awaiting-merge 

---

Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6355)

**Pull request merged**